File: integrations/gmail/gmail_send_email.go
package gmail

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"mime"
	"net/http"
	"path"
	"strings"
	"time"

	"git.sr.ht/~klahr/hazy-flow/core"
	"git.sr.ht/~klahr/hazy-flow/engine"
	"git.sr.ht/~klahr/hazy-flow/integrations/internal/params"
	"git.sr.ht/~klahr/hazy-flow/integrations/internal/sandbox"
)

func init() {
	engine.Register(engine.NativeDrop{
		Manifest: core.Manifest{
			ID:             "gmail_send_email",
			Version:        "1.1",
			Label:          "Gmail send email",
			Color:          "#D14836",
			Icon:           "mail",
			BrandLogo:      "/brands/gmail.svg",
			Category:       "network",
			Provider:       "internal",
			Integration:    "Gmail",
			Tags:           []string{"gmail", "email", "send", "google", "attachments"},
			Description: "Send an email through your connected Gmail account. The body comes from the 'body' input or from the 'body' param; set format to html for rich content. Wire any number of file-producing nodes into the 'attachments' input port to attach files (PDFs, spreadsheets, anything else) — each ref's MIME + filename ride through to the recipient. The 'from' address is fixed to your authorized Google account — Gmail's API doesn't allow spoofing.",
			Summary:     "Send an email through the user's connected Gmail account, in plain text or HTML, with optional attachments and threading.",
			Examples: []core.ParamsExample{
				{
					Title:  "Plain-text alert",
					Params: json.RawMessage(`{"to":"oncall@example.com","subject":"Build failed","body":"main is red, see https://ci.example.com/run/123","token":"${secret:GMAIL_OAUTH}"}`),
				},
				{
					Title:  "HTML newsletter to a list",
					Params: json.RawMessage(`{"to":"team@example.com","cc":"leads@example.com","subject":"Weekly digest","body":"<h1>Highlights</h1><p>Shipped the new onboarding flow.</p>","format":"html","token":"${secret:GMAIL_OAUTH}"}`),
				},
				{
					Title:  "Threaded reply",
					Params: json.RawMessage(`{"to":"alice@example.com","subject":"Re: deploy plan","body":"Sounds good — running it at 16:00 UTC.","thread_id":"18f9d3a2c0e1b4a5","reply_to":"ops@example.com","token":"${secret:GMAIL_OAUTH}"}`),
					Notes:  "thread_id keeps the reply in the same Gmail conversation as the original message.",
				},
				{
					Title:  "Daily report with the spreadsheet as a PDF attachment",
					Params: json.RawMessage(`{"to":"me@example.com","subject":"Yesterday's comments","body":"Comments digest attached.","token":"${secret:GMAIL_OAUTH}"}`),
					Notes:  "Wire a sheets_export_pdf (or file_write of a generated PDF) output into the 'attachments' input port. The port is variadic — connect as many file-producing nodes as you want and each becomes its own attachment.",
				},
			},
			RequiresConnections: []core.ConnectionRequirement{
				{Kind: "oauth", Name: "gmail", Note: "Gmail OAuth — gmail.send / gmail.readonly scopes."},
			},
			ExecutionModel: core.ExecutionBatch,
			ProcessModel:   core.ProcessLongLived,
			Inputs: []core.Port{
				{Port: "body", Label: "Email body (overrides params.body)"},
				{
					Port:     "attachments",
					Label:    "Files to attach (wire zero or more file-producing nodes here)",
					Variadic: true,
				},
			},
			Outputs: []core.Port{
				{Port: "meta", Label: "Delivery metadata", MIME: []string{"application/json"}},
			},
			ParamsSchema: json.RawMessage(`{
				"type":"object",
				"properties":{
					"account":   {"type":"string","default":"default"},
					"token":     {"type":"string","description":"Raw access token; overrides 'account'."},
					"to":        {"type":"string","description":"Recipient address (or comma-separated list)."},
					"cc":        {"type":"string"},
					"bcc":       {"type":"string"},
					"subject":   {"type":"string"},
					"body":      {"type":"string","description":"Default body when the input port isn't wired."},
					"format":    {"type":"string","enum":["text","html"],"default":"text","description":"text/plain vs text/html Content-Type."},
					"reply_to":  {"type":"string","description":"Reply-To header."},
					"thread_id":{"type":"string","description":"Gmail thread ID to thread this reply into."},
					"timeout_ms":{"type":"integer","default":15000,"minimum":1}
				},
				"required":["to"]
			}`),
			Idempotent:  false,
			RetryPolicy: core.RetryExponentialBackoff,
		},
		Execute: executeGmailSendEmail,
	})
}

// executeGmailSendEmail constructs an RFC822 message, base64-URL-
// encodes it (the Gmail API's required wire format — NOT std
// base64, the URL variant without padding), and POSTs to
// users/me/messages/send. The "me" alias means "the authorized
// user", so the sender is implicitly the OAuth-connected account.
func executeGmailSendEmail(ctx context.Context, job core.Job, _ chan<- core.Progress) (core.Result, error) {
	to, err := params.String(job.Params, "to")
	if err != nil {
		return params.Err(job, "bad_param", err.Error()), nil
	}
	token, err := resolveToken(ctx, job)
	if err != nil {
		return params.Err(job, "auth", err.Error()), nil
	}

	body, _ := params.StringOpt(job.Params, "body")
	if input, ok := job.Input["body"]; ok && input.Inline != nil {
		switch v := input.Inline.(type) {
		case string:
			body = v
		case []byte:
			body = string(v)
		default:
			return params.Err(job, "bad_input",
				fmt.Sprintf("body: expected string (text or HTML); got %T", v)), nil
		}
	}
	if body == "" {
		return params.Err(job, "bad_input", "no body — set params.body or wire the 'body' input port"), nil
	}

	subject := params.StringDefault(job.Params, "subject", "(no subject)")
	format := params.StringDefault(job.Params, "format", "text")
	bodyContentType := "text/plain; charset=\"utf-8\""
	if format == "html" {
		bodyContentType = "text/html; charset=\"utf-8\""
	}

	attachments, err := loadAttachments(job)
	if err != nil {
		return params.Err(job, "bad_input", err.Error()), nil
	}

	raw, err := buildRFC822(rfc822Headers{
		To:              to,
		Cc:              params.StringDefault(job.Params, "cc", ""),
		Bcc:             params.StringDefault(job.Params, "bcc", ""),
		Subject:         subject,
		ReplyTo:         params.StringDefault(job.Params, "reply_to", ""),
		BodyContentType: bodyContentType,
	}, body, attachments)
	if err != nil {
		return params.Err(job, "internal", fmt.Sprintf("build rfc822: %v", err)), nil
	}

	payload := map[string]any{
		// base64-URL-encode without padding — Gmail's required
		// wire format for raw messages.
		"raw": base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString(raw),
	}
	if tid, ok := params.StringOpt(job.Params, "thread_id"); ok && tid != "" {
		payload["threadId"] = tid
	}
	jsonBody, _ := json.Marshal(payload)

	url := currentHTTPBase() + "/users/me/messages/send"
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(jsonBody))
	if err != nil {
		return params.Err(job, "internal", err.Error()), nil
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json; charset=utf-8")
	// Idempotency-Key prevents double-send on worker retry. Gmail
	// itself doesn't fully honor the header today, but sending it
	// is harmless to APIs that ignore it and forward-compatible
	// if Google starts deduping (the Cloud APIs they front do).
	req.Header.Set("Idempotency-Key", job.IdempotencyKey())

	timeoutMs := params.IntDefault(job.Params, "timeout_ms", 15000)
	client := &http.Client{Timeout: time.Duration(timeoutMs) * time.Millisecond}
	resp, err := client.Do(req)
	if err != nil {
		return params.Err(job, "send_failed", err.Error()), nil
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(io.LimitReader(resp.Body, 64*1024))

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return params.Err(job, "gmail_error",
			fmt.Sprintf("Gmail returned %d: %s", resp.StatusCode, extractGmailError(respBody))), nil
	}

	var parsed map[string]any
	_ = json.Unmarshal(respBody, &parsed)
	meta := map[string]any{
		"id":       stringField(parsed, "id"),
		"threadId": stringField(parsed, "threadId"),
	}
	return core.Result{
		JobID:  job.ID,
		Status: core.StatusOK,
		Output: map[string]core.Ref{
			"meta": {MIME: "application/json", Inline: meta},
		},
	}, nil
}

// rfc822Headers is the small subset of email headers we care about.
// BodyContentType is the inner body part's type ("text/plain" or
// "text/html"); the outer Content-Type is computed by buildRFC822 and
// flips to multipart/mixed when attachments are present.
type rfc822Headers struct {
	To              string
	Cc              string
	Bcc             string
	Subject         string
	ReplyTo         string
	BodyContentType string
}

// emailAttachment carries one file's worth of bytes plus the metadata
// the recipient needs to render it (filename for Content-Disposition,
// MIME type for Content-Type). Resolved from a variadic input port at
// run time — see loadAttachments.
type emailAttachment struct {
	Filename string
	MIME     string
	Bytes    []byte
}

// buildRFC822 builds the wire-format message Gmail expects. CRLF line
// endings are spec-mandated; lone-LF works for many MTAs but some
// choke. With no attachments we emit a single-part message (preserving
// the historical wire format); with one or more we emit multipart/mixed
// with the body as the first part and each attachment as a subsequent
// base64-encoded part.
func buildRFC822(h rfc822Headers, body string, attachments []emailAttachment) ([]byte, error) {
	var buf bytes.Buffer
	headerInto := func(w *bytes.Buffer, name, value string) {
		if value == "" {
			return
		}
		// Strip CR/LF from header values — header injection defense.
		// A user supplying "test@x\r\nBcc: leak@y" must not be able to
		// add headers.
		v := strings.NewReplacer("\r", "", "\n", "").Replace(value)
		fmt.Fprintf(w, "%s: %s\r\n", name, v)
	}
	headerInto(&buf, "To", h.To)
	headerInto(&buf, "Cc", h.Cc)
	headerInto(&buf, "Bcc", h.Bcc)
	headerInto(&buf, "Reply-To", h.ReplyTo)
	// RFC 2047 encoded-word: email headers are 7-bit ASCII, so a subject
	// with non-ASCII (e.g. Swedish "Hej från Hazy Flow!") must be encoded
	// as =?utf-8?q?...?= rather than shipped as raw UTF-8 bytes — raw bytes
	// get mojibake'd by receiving clients. QEncoding leaves pure-ASCII
	// subjects untouched, so this is a no-op for the common case.
	headerInto(&buf, "Subject", mime.QEncoding.Encode("utf-8", h.Subject))
	headerInto(&buf, "MIME-Version", "1.0")

	if len(attachments) == 0 {
		headerInto(&buf, "Content-Type", h.BodyContentType)
		headerInto(&buf, "Content-Transfer-Encoding", "8bit")
		buf.WriteString("\r\n")
		buf.WriteString(body)
		return buf.Bytes(), nil
	}

	boundary, err := newBoundary()
	if err != nil {
		return nil, err
	}
	headerInto(&buf, "Content-Type", fmt.Sprintf("multipart/mixed; boundary=%q", boundary))
	buf.WriteString("\r\n")

	// Body part.
	fmt.Fprintf(&buf, "--%s\r\n", boundary)
	headerInto(&buf, "Content-Type", h.BodyContentType)
	headerInto(&buf, "Content-Transfer-Encoding", "8bit")
	buf.WriteString("\r\n")
	buf.WriteString(body)
	buf.WriteString("\r\n")

	// Attachment parts, base64-encoded with 76-column line wraps (the
	// classic MIME quoted-line width). Some receivers are strict about
	// the wrap; stdlib base64 stops short of producing that for us so
	// we wrap manually.
	for _, a := range attachments {
		fmt.Fprintf(&buf, "--%s\r\n", boundary)
		headerInto(&buf, "Content-Type", a.MIME)
		headerInto(&buf, "Content-Disposition", dispositionHeader(a.Filename))
		headerInto(&buf, "Content-Transfer-Encoding", "base64")
		buf.WriteString("\r\n")
		encoded := base64.StdEncoding.EncodeToString(a.Bytes)
		for i := 0; i < len(encoded); i += 76 {
			end := i + 76
			if end > len(encoded) {
				end = len(encoded)
			}
			buf.WriteString(encoded[i:end])
			buf.WriteString("\r\n")
		}
	}
	fmt.Fprintf(&buf, "--%s--\r\n", boundary)
	return buf.Bytes(), nil
}

// dispositionHeader builds the Content-Disposition value for an
// attachment. A non-ASCII filename (e.g. "årsrapport.pdf") can't ride as
// a raw header value, so mime.FormatMediaType emits the RFC 2231 form
// (filename*=utf-8''%C3%A5rsrapport.pdf); ASCII names stay as the plain
// filename= form. FormatMediaType returns "" if it can't encode the
// params — fall back to a quoted plain filename so we still send something
// sane rather than an empty header.
func dispositionHeader(filename string) string {
	if v := mime.FormatMediaType("attachment", map[string]string{"filename": filename}); v != "" {
		return v
	}
	return fmt.Sprintf("attachment; filename=%q", filename)
}

// newBoundary returns a MIME boundary token unlikely to appear inside
// any attachment body. 16 random bytes hex-encoded gives 32 hex chars
// of entropy — collision-free for any practical message.
func newBoundary() (string, error) {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", fmt.Errorf("boundary entropy: %w", err)
	}
	return "hazyflow-" + hex.EncodeToString(b[:]), nil
}

// loadAttachments collects every Ref on the variadic "attachments"
// input port and materialises the bytes. A Ref carries either inline
// bytes (string / []byte) or a sandbox path; either is resolved to a
// concrete []byte the MIME builder can base64-encode.
//
// Filename precedence: an explicit basename from the Ref's sandbox
// path beats a synthesised "attachment-N" — keeps "report.pdf" as
// "report.pdf" in the recipient's inbox.
func loadAttachments(job core.Job) ([]emailAttachment, error) {
	refs := core.VariadicInputs(job.Input, "attachments")
	if len(refs) == 0 {
		return nil, nil
	}
	out := make([]emailAttachment, 0, len(refs))
	for i, ref := range refs {
		data, err := readAttachmentBytes(job, ref)
		if err != nil {
			return nil, fmt.Errorf("attachment %d: %w", i, err)
		}
		out = append(out, emailAttachment{
			Filename: attachmentFilename(ref, i),
			MIME:     attachmentMIME(ref),
			Bytes:    data,
		})
	}
	return out, nil
}

// readAttachmentBytes reads the bytes a Ref points at. Inline strings/
// []byte are taken verbatim; a sandbox path (workspace:// implicit, or
// scratch://) is opened via the shared sandbox helper that confines
// reads to the per-run roots.
func readAttachmentBytes(job core.Job, ref core.Ref) ([]byte, error) {
	if ref.Inline != nil {
		switch v := ref.Inline.(type) {
		case []byte:
			return v, nil
		case string:
			return []byte(v), nil
		default:
			return nil, fmt.Errorf("inline attachment must be bytes or string, got %T", v)
		}
	}
	if ref.Ref == "" {
		return nil, fmt.Errorf("ref has neither inline bytes nor a sandbox path")
	}
	root, rel, err := sandbox.OpenRoot(job, ref.Ref)
	if err != nil {
		return nil, err
	}
	defer root.Close()
	f, err := root.Open(rel)
	if err != nil {
		if sandbox.IsEscape(err) {
			return nil, fmt.Errorf("path %q escapes its sandbox root", ref.Ref)
		}
		return nil, fmt.Errorf("open %q: %w", ref.Ref, err)
	}
	defer f.Close()
	return io.ReadAll(f)
}

// attachmentFilename picks a sensible filename for the
// Content-Disposition header. Sandbox path basename wins (it's what
// the user named the file); inline refs get a synthesised name with an
// extension guessed from the MIME, falling back to .bin.
func attachmentFilename(ref core.Ref, idx int) string {
	if ref.Ref != "" {
		if name := path.Base(ref.Ref); name != "" && name != "." && name != "/" {
			return name
		}
	}
	return fmt.Sprintf("attachment-%d%s", idx+1, extForMIME(ref.MIME))
}

// attachmentMIME returns the declared MIME or a generic-binary fallback
// the receiver can still save and inspect.
func attachmentMIME(ref core.Ref) string {
	if ref.MIME != "" {
		return ref.MIME
	}
	return "application/octet-stream"
}

// extForMIME maps the handful of MIME types Hazy Flow file-producing
// drops emit today to their canonical extension. Anything outside the
// list lands as .bin — better than no extension, and the MIME header is
// the authoritative signal anyway.
func extForMIME(mime string) string {
	switch strings.ToLower(strings.SplitN(mime, ";", 2)[0]) {
	case "application/pdf":
		return ".pdf"
	case "text/plain":
		return ".txt"
	case "text/csv":
		return ".csv"
	case "text/html":
		return ".html"
	case "application/json":
		return ".json"
	case "image/png":
		return ".png"
	case "image/jpeg":
		return ".jpg"
	case "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet":
		return ".xlsx"
	case "application/zip":
		return ".zip"
	}
	return ".bin"
}

func stringField(m map[string]any, k string) string {
	if m == nil {
		return ""
	}
	if s, ok := m[k].(string); ok {
		return s
	}
	return ""
}

File: officialdrops/gmail_send_email.ts
/**
 * gmail_send_email — the pilot port of a production connector to a scripted drop.
 *
 * Behaviour-for-behaviour with the native Go drop it replaces: builds an RFC822
 * message (header-injection-safe, RFC 2047 encoded-word subject so non-ASCII
 * like "Hej från Hazyflow!" doesn't mojibake, multipart/mixed with base64
 * attachments + RFC 2231 filenames), base64url-encodes it the way Gmail's API
 * requires, and POSTs to users/me/messages/send. The OAuth-connected account is
 * the implicit sender ("me"); Gmail's API doesn't allow spoofing the From.
 *
 * It reaches the network only through ctx.fetch (SSRF-guarded), the token only
 * through ctx.auth, attachments only through ctx.inputs + ctx.files — the same
 * capability surface that becomes the container broker API later. No native Go.
 */

const GMAIL_API_BASE = "https://gmail.googleapis.com/gmail/v1";

export default {
  manifest: {
    id: "gmail_send_email",
    version: "2.0.0",
    label: "Gmail send email",
    summary:
      "Send an email through the user's connected Gmail account, in plain text or HTML, with optional attachments and threading.",
    description:
      "Send an email through your connected Gmail account. The body comes from the 'body' input or the 'body' param; set format to html for rich content. Wire any number of file-producing nodes into the variadic 'attachments' input to attach files — each ref's MIME + filename ride through to the recipient. The 'from' address is fixed to your authorized Google account.",
    integration: "Gmail",
    category: "network",
    icon: "mail",
    brandLogo: "/brands/gmail.svg",
    color: "#D14836",
    tags: ["gmail", "email", "send", "google", "attachments"],
    requiresConnections: [
      { kind: "oauth", name: "google", note: "Google OAuth — gmail.send scope." },
    ],
    inputs: [
      { port: "body", label: "Email body (overrides params.body)" },
      { port: "attachments", label: "Files to attach (wire zero or more file-producing nodes)", variadic: true },
    ],
    outputs: [{ port: "meta", label: "Delivery metadata", mime: ["application/json"] }],
    idempotent: false,
    retryPolicy: "exponential_backoff",
    paramsSchema: {
      type: "object",
      properties: {
        base_url: { type: "string", description: "Override the API host (proxy / self-hosted / testing)." },
        account: { type: "string", default: "default" },
        token: { type: "string", description: "Raw access token; overrides 'account'." },
        to: { type: "string", description: "Recipient address (or comma-separated list)." },
        cc: { type: "string" },
        bcc: { type: "string" },
        subject: { type: "string" },
        body: { type: "string", description: "Default body when the input port isn't wired." },
        format: { type: "string", enum: ["text", "html"], default: "text" },
        reply_to: { type: "string", description: "Reply-To header." },
        thread_id: { type: "string", description: "Gmail thread ID to thread this reply into." },
        timeout_ms: { type: "integer", default: 15000, minimum: 1 },
      },
      required: ["to"],
    },
    examples: [
      {
        title: "Plain-text alert",
        params: { to: "oncall@example.com", subject: "Build failed", body: "main is red", token: "${secret:GMAIL_OAUTH}" },
      },
      {
        title: "HTML newsletter to a list",
        params: { to: "team@example.com", cc: "leads@example.com", subject: "Weekly digest", body: "<h1>Highlights</h1>", format: "html", token: "${secret:GMAIL_OAUTH}" },
      },
      {
        title: "Daily report with a PDF attachment",
        params: { to: "me@example.com", subject: "Yesterday's comments", body: "Comments digest attached.", token: "${secret:GMAIL_OAUTH}" },
        notes: "Wire a file-producing node (e.g. sheets_export_pdf) into the variadic 'attachments' input.",
      },
    ],
  },

  async run(ctx: any) {
    const p = ctx.params || {};
    const base = String(p.base_url || GMAIL_API_BASE).replace(/\/+$/, "");

    const to = String(p.to || "").trim();
    if (!to) throw new DropError("bad_param", "'to' is required");

    // Token: an explicit param wins (covers ${secret:…} templating); otherwise
    // the host-refreshed OAuth token for the gmail integration.
    let token = String(p.token || "").trim();
    if (!token) {
      if (!ctx.auth) throw new DropError("auth", "no token supplied and OAuth is not configured");
      token = await ctx.auth.token("google", p.account || "default");
    }
    if (!token) throw new DropError("auth", "empty access token");

    // Body: the input port overrides params.body. A wired ref is inline text or
    // a file-backed path we read through the sandbox.
    let body = String(p.body || "");
    if (ctx.inputs.has("body")) {
      const ref = ctx.inputs.ref("body");
      if (typeof ref.value === "string") body = ref.value;
      else if (ref.path) body = await ctx.files.readText(ref.path);
      else throw new DropError("bad_input", "body input must be text");
    }
    if (!body) throw new DropError("bad_input", "no body — set params.body or wire the 'body' input port");

    const bodyContentType =
      p.format === "html" ? 'text/html; charset="utf-8"' : 'text/plain; charset="utf-8"';

    const attachments = await loadAttachments(ctx);

    const raw = buildRFC822(
      ctx,
      {
        to,
        cc: String(p.cc || ""),
        bcc: String(p.bcc || ""),
        subject: String(p.subject || "(no subject)"),
        replyTo: String(p.reply_to || ""),
        bodyContentType,
      },
      body,
      attachments,
    );

    const payload: any = {
      // Gmail requires base64url WITHOUT padding for the raw message.
      raw: ctx.crypto.base64(ctx.crypto.utf8(raw), { url: true, pad: false }),
    };
    if (p.thread_id) payload.threadId = String(p.thread_id);

    const res = await ctx.fetch(`${base}/users/me/messages/send`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json; charset=utf-8",
      },
      body: JSON.stringify(payload),
      timeoutMs: Number(p.timeout_ms) || 15000,
    });

    if (!res.ok) {
      const errText = await res.text();
      throw new DropError("gmail_error", `Gmail returned ${res.status}: ${extractGmailError(errText)}`);
    }

    const parsed: any = await res.json();
    return { meta: { id: parsed.id || "", threadId: parsed.threadId || "" } };
  },
};

// ───────────────────────────────────────────────────────────── attachments ──

interface Attachment {
  filename: string;
  mime: string;
  base64: string; // standard base64, already wrapped at 76 columns
}

async function loadAttachments(ctx: any): Promise<Attachment[]> {
  const refs = ctx.inputs.all("attachments");
  const out: Attachment[] = [];
  for (let i = 0; i < refs.length; i++) {
    const r = refs[i];
    const bytes =
      r.value !== undefined && r.value !== null ? r.value : await ctx.files.read(r.path);
    out.push({
      filename: attachmentFilename(r, i),
      mime: r.mime || "application/octet-stream",
      base64: wrap76(ctx.crypto.base64(bytes)), // std base64, padded
    });
  }
  return out;
}

function attachmentFilename(ref: any, idx: number): string {
  if (ref.path) {
    const base = ref.path.split("/").pop();
    if (base && base !== "." ) return base;
  }
  return `attachment-${idx + 1}${extForMIME(ref.mime || "")}`;
}

function extForMIME(mime: string): string {
  switch (mime.split(";")[0].trim().toLowerCase()) {
    case "application/pdf": return ".pdf";
    case "text/plain": return ".txt";
    case "text/csv": return ".csv";
    case "text/html": return ".html";
    case "application/json": return ".json";
    case "image/png": return ".png";
    case "image/jpeg": return ".jpg";
    case "application/zip": return ".zip";
  }
  return ".bin";
}

// ─────────────────────────────────────────────────────────── RFC822 builder ──

interface Headers {
  to: string;
  cc: string;
  bcc: string;
  subject: string;
  replyTo: string;
  bodyContentType: string;
}

function buildRFC822(ctx: any, h: Headers, body: string, attachments: Attachment[]): string {
  const lines: string[] = [];
  const header = (name: string, value: string) => {
    if (!value) return;
    // Header-injection defense: a value like "x@y\r\nBcc: leak@z" must not
    // smuggle extra headers.
    lines.push(`${name}: ${value.replace(/[\r\n]/g, "")}`);
  };

  header("To", h.to);
  header("Cc", h.cc);
  header("Bcc", h.bcc);
  header("Reply-To", h.replyTo);
  header("Subject", encodeWord(ctx, h.subject));
  header("MIME-Version", "1.0");

  if (attachments.length === 0) {
    header("Content-Type", h.bodyContentType);
    header("Content-Transfer-Encoding", "8bit");
    return lines.join("\r\n") + "\r\n\r\n" + body;
  }

  const boundary = "hazyflow-" + ctx.crypto.hex(ctx.crypto.randomBytes(16));
  header("Content-Type", `multipart/mixed; boundary="${boundary}"`);

  const parts: string[] = [lines.join("\r\n") + "\r\n"];

  // Body part.
  parts.push(
    `--${boundary}\r\n` +
      `Content-Type: ${h.bodyContentType}\r\n` +
      `Content-Transfer-Encoding: 8bit\r\n\r\n` +
      body +
      "\r\n",
  );

  // Attachment parts.
  for (const a of attachments) {
    parts.push(
      `--${boundary}\r\n` +
        `Content-Type: ${a.mime.replace(/[\r\n]/g, "")}\r\n` +
        `Content-Disposition: ${dispositionHeader(ctx, a.filename)}\r\n` +
        `Content-Transfer-Encoding: base64\r\n\r\n` +
        a.base64 +
        "\r\n",
    );
  }
  parts.push(`--${boundary}--\r\n`);
  return parts.join("");
}

// encodeWord applies RFC 2047 Q-encoding when a subject contains non-ASCII, and
// leaves pure-ASCII subjects untouched (mirrors Go's mime.QEncoding).
function encodeWord(ctx: any, s: string): string {
  if (isAscii(s)) return s;
  const bytes = utf8Bytes(s);
  let enc = "";
  for (const b of bytes) {
    if (b === 0x20) enc += "_";
    else if (b >= 0x21 && b <= 0x7e && b !== 0x3d && b !== 0x3f && b !== 0x5f)
      enc += String.fromCharCode(b);
    // RFC 2045 quoted-printable mandates uppercase hex digits (=C3=A5),
    // matching the native drop's mime.QEncoding output.
    else enc += "=" + hex2(b).toUpperCase();
  }
  return `=?utf-8?q?${enc}?=`;
}

// dispositionHeader emits a plain quoted filename for ASCII names and the RFC
// 2231 extended form (filename*=utf-8''…) for non-ASCII (mirrors Go's
// mime.FormatMediaType), so "årsrapport.pdf" survives to the recipient.
function dispositionHeader(_ctx: any, filename: string): string {
  if (isAscii(filename)) {
    return `attachment; filename="${filename.replace(/[\r\n"]/g, "")}"`;
  }
  let pct = "";
  for (const b of utf8Bytes(filename)) {
    if (isAttrChar(b)) pct += String.fromCharCode(b);
    else pct += "%" + hex2(b).toUpperCase();
  }
  return `attachment; filename*=utf-8''${pct}`;
}

// ───────────────────────────────────────────────────────────────── helpers ──

function isAscii(s: string): boolean {
  for (let i = 0; i < s.length; i++) if (s.charCodeAt(i) > 127) return false;
  return true;
}

// utf8Bytes encodes a JS string to UTF-8 as a plain number array — no runtime
// TextEncoder/Uint8Array iteration assumptions.
function utf8Bytes(s: string): number[] {
  const out: number[] = [];
  for (let i = 0; i < s.length; i++) {
    let c = s.charCodeAt(i);
    if (c < 0x80) out.push(c);
    else if (c < 0x800) out.push(0xc0 | (c >> 6), 0x80 | (c & 0x3f));
    else if (c >= 0xd800 && c <= 0xdbff) {
      const c2 = s.charCodeAt(++i);
      const cp = 0x10000 + ((c - 0xd800) << 10) + (c2 - 0xdc00);
      out.push(0xf0 | (cp >> 18), 0x80 | ((cp >> 12) & 0x3f), 0x80 | ((cp >> 6) & 0x3f), 0x80 | (cp & 0x3f));
    } else out.push(0xe0 | (c >> 12), 0x80 | ((c >> 6) & 0x3f), 0x80 | (c & 0x3f));
  }
  return out;
}

function hex2(b: number): string {
  return (b < 16 ? "0" : "") + b.toString(16);
}

// RFC 2231 attr-char set (the bytes that may appear unencoded in filename*).
function isAttrChar(b: number): boolean {
  if ((b >= 0x30 && b <= 0x39) || (b >= 0x41 && b <= 0x5a) || (b >= 0x61 && b <= 0x7a)) return true;
  return "!#$&+-.^_`|~".indexOf(String.fromCharCode(b)) >= 0;
}

function wrap76(b64: string): string {
  const rows: string[] = [];
  for (let i = 0; i < b64.length; i += 76) rows.push(b64.slice(i, i + 76));
  return rows.join("\r\n");
}

function extractGmailError(body: string): string {
  try {
    const j = JSON.parse(body);
    if (j && j.error && j.error.message) return String(j.error.message);
  } catch (_e) {
    // not JSON — fall through
  }
  return body.slice(0, 200);
}
